lib/helpers/wp-cli-db/class-db-server.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace Automattic\VIP\Helpers\WP_CLI_DB;
+
+class DB_Server {
+	private $host;
+	private $user;
+	private $password;
+	private $name;
+	private $read_priority;
+	private $write_priority;
+
+	public function __construct( string $host, string $user, string $password, string $name, int $read_priority, int $write_priority ) {
+		$this->host           = $host;
+		$this->user           = $user;
+		$this->password       = $password;
+		$this->name           = $name;
+		$this->read_priority  = $read_priority;
+		$this->write_priority = $write_priority;
+	}
+
+	public function define_variables() {
+		if ( ! defined( 'DB_HOST' ) ) {
+			define( 'DB_HOST', $this->host );
+		}
+		if ( ! defined( 'DB_USER' ) ) {
+			define( 'DB_USER', $this->user );
+		}
+		if ( ! defined( 'DB_PASSWORD' ) ) {
+			define( 'DB_PASSWORD', $this->password );
+		}
+		if ( ! defined( 'DB_NAME' ) ) {
+			define( 'DB_NAME', $this->name );
+		}
```

Comment:
Thanks much for taking a look, @WPprodigy!

> Do we have to use the "main" constants

It might be possible to override the arguments that get passed to the WP-CLI invocation, but it's not immediately apparent how to do it:

https://github.com/wp-cli/db-command/blob/de3eddeaa224e617a05fb3a8a97d77fefc456c4f/src/DB_Command.php#L1595-L1597
https://github.com/wp-cli/db-command/blob/de3eddeaa224e617a05fb3a8a97d77fefc456c4f/src/DB_Command.php#L453

> This is something that might interfere with our other WP bootup logic in the future.

These constants are only going to be set in the context of the `wp db` scope / lifetime and WordPress never loads in that context, so it oughtn't affect other things.

> We're initializing multiple DB_Servers.

This code intentionally chooses a single server from the list (if any suitable ones are present in the global) to fulfill the command.

I feel like I'm missing something from your point `B`, though. Can you expand, please?

Reply:
Okay I see now. there are multiple initializations of `DB_Server` happening here: https://github.com/Automattic/vip-go-mu-plugins/pull/3064/files#diff-a182b0d077b2b93f40cb08b85e87943fea9086cac0b60c145c93215bfa636148R55-R63

But eventually is filtered down to just one, and only that one has `define_variables()` called on it.